aws/amazon-ecs-agent
# save and load pulled containers state

### Summary
This PR is a continuation of https://github.com/aws/amazon-ecs-agent/pull/2716 . So please review only the 2nd commit. 

- saves and loads the new pulled containers state to and from agent's boltdb state. 
- It also removes it from state when container moves to `CREATED` state. The removal is also attempted when task is stopped, in case containers didn't move to `CREATED` before moving to `STOPPED` state.
- updates the already added pulled container state when known status of container is updated to `PULLED`

### Implementation details
- Add a new bucket named `pulledcontainers` in boltdb state 
- Introduce methods `SavePulledContainer`, `DeletePulledContainer`, `GetPulledContainers` in data client for saving, removing and loading data from boltdb
- update the already added pulled container state when known status of container is updated to `PULLED` in `managedTask.handleContainerChange` method.
- today pulled containers are added to containers table in agent state here https://github.com/aws/amazon-ecs-agent/blob/1489adfaec05c559ea4e1490774d1f86261cf936/agent/engine/task_manager.go#L422 As a result, when agent restarts, the pulled container is added to non-pulled containers state, which then is attempted to obtain network metadata here https://github.com/aws/amazon-ecs-agent/blob/84321fcb15787f1d3a3f8aed80002d6d7d080ec6/agent/handlers/v4/container_metadata_handler.go#L74 and fails here https://github.com/aws/amazon-ecs-agent/blob/84321fcb15787f1d3a3f8aed80002d6d7d080ec6/agent/handlers/v4/container_metadata_handler.go#L85 
Fixed that in this PR, by skipping if container's known states is `PULLED` and container ID is empty. 

### Testing



New tests cover the changes: 
- Unit tests yet to be updated 
- Manual testing done: 
TMDE V4 response:
```
curl ${ECS_CONTAINER_METADATA_URI_V4}/task
{"Cluster":"pull-state","TaskARN":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","Family":"dep-check","Revision":"1","DesiredStatus":"RUNNING","KnownStatus":"NONE","PullStartedAt":"2020-11-11T14:56:03.500537906Z","PullStoppedAt":"2020-11-11T14:56:04.87932092Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"DockerId":"5e733312cebf2b606e9a41af031f42a596d5a025ae34a407d359c9a6d95b7f30","Name":"network","DockerName":"ecs-dep-check-1-network-daebabb79381aea78901","Image":"ubuntu","ImageID":"sha256:d70eaf7277eada08fca944de400e7e4dd97b1262c06ed2b1011500caa4decaf1","Labels":{"com.amazonaws.ecs.cluster":"pull-state","com.amazonaws.ecs.container-name":"network","com.amazonaws.ecs.task-arn":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","com.amazonaws.ecs.task-definition-family":"dep-check","com.amazonaws.ecs.task-definition-version":"1"},"DesiredStatus":"RUNNING","KnownStatus":"RUNNING","Limits":{"CPU":0,"Memory":512},"CreatedAt":"2020-11-11T14:56:04.777143666Z","StartedAt":"2020-11-11T14:56:05.421289274Z","Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/fb246e8e-4dab-4c3a-b7f6-f753c1ee1b30","Networks":[{"NetworkMode":"bridge","IPv4Addresses":["172.17.0.2"]}]},{"DockerId":"","Name":"customer","DockerName":"","Image":"amazonlinux:latest","ImageID":"sha256:ba2cc467a2bce7ffc9e2d6e38a87f6ec89c1c93bc83ff575e8021bb592f9ad2e","DesiredStatus":"RUNNING","KnownStatus":"PULLED","Limits":{"CPU":0,"Memory":512},"Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/7a400635-6cd2-4158-974d-bfab9885abab"}]} 
``` 
Restart agent now 
TMDE response after agent restart (pulled container is still seen)
```
{"Cluster":"pull-state","TaskARN":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","Family":"dep-check","Revision":"1","DesiredStatus":"RUNNING","KnownStatus":"NONE","PullStartedAt":"2020-11-11T14:56:03.500537906Z","PullStoppedAt":"2020-11-11T14:56:04.87932092Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"DockerId":"5e733312cebf2b606e9a41af031f42a596d5a025ae34a407d359c9a6d95b7f30","Name":"network","DockerName":"ecs-dep-check-1-network-daebabb79381aea78901","Image":"ubuntu","ImageID":"sha256:d70eaf7277eada08fca944de400e7e4dd97b1262c06ed2b1011500caa4decaf1","Labels":{"com.amazonaws.ecs.cluster":"pull-state","com.amazonaws.ecs.container-name":"network","com.amazonaws.ecs.task-arn":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","com.amazonaws.ecs.task-definition-family":"dep-check","com.amazonaws.ecs.task-definition-version":"1"},"DesiredStatus":"RUNNING","KnownStatus":"RUNNING","Limits":{"CPU":0,"Memory":512},"CreatedAt":"2020-11-11T14:56:04.777143666Z","StartedAt":"2020-11-11T14:56:05.421289274Z","Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/fb246e8e-4dab-4c3a-b7f6-f753c1ee1b30","Networks":[{"NetworkMode":"bridge","IPv4Addresses":["172.17.0.2"]}]},{"DockerId":"","Name":"customer","DockerName":"","Image":"amazonlinux:latest","ImageID":"sha256:ba2cc467a2bce7ffc9e2d6e38a87f6ec89c1c93bc83ff575e8021bb592f9ad2e","DesiredStatus":"RUNNING","KnownStatus":"PULLED","Limits":{"CPU":0,"Memory":512},"Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/7a400635-6cd2-4158-974d-bfab9885abab"}]}
```
example of `agent.db` with pulled container added:
```
| Path: pulledcontainers → ea86e60a718745eb850fd6afefa4eb1f-customer
| Key: ea86e60a718745eb850fd6afefa4eb1f-customer
| Value:
| {
|   "Container": {
|     "ApplyingError": null,
|     "Command": [
|       "sleep 100"
|     ],
|     "ContainerArn": "arn:aws:ecs:us-west-2:074626758757:contai....
….
….
|     "KnownStatus": "PULLED",
```
once container moves to created:
```
| Path: pulledcontainers
| Buckets: 0
| Pairs: 0
```

### Description for the changelog


### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
